sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,516 @@
+        String queue,
+        String topic) {
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+    }
+
+    @Override
+    public List<UnboundedJmsSource> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      List<UnboundedJmsSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        sources.add(new UnboundedJmsSource(connectionFactory, queue, topic));
+      }
+      return sources;
+    }
+
+    @Override
+    public UnboundedJmsReader createReader(PipelineOptions options,
+                                           JmsCheckpointMark checkpointMark) {
+      return new UnboundedJmsReader(this, checkpointMark);
+    }
+
+    @Override
+    public void validate() {
+      checkNotNull(connectionFactory, "ConnectionFactory is not defined");
+      checkArgument((queue != null || topic != null), "Either queue or topic is not defined");
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return AvroCoder.of(JmsCheckpointMark.class);
+    }
+
+    @Override
+    public Coder<JmsRecord> getDefaultOutputCoder() {
+      return SerializableCoder.of(JmsRecord.class);
+    }
+
+  }
+
+  private static class UnboundedJmsReader extends UnboundedReader<JmsRecord> {
+
+    private UnboundedJmsSource source;
+    private JmsCheckpointMark checkpointMark;
+    private String destination;
+    private Connection connection;
+    private Session session;
+    private MessageConsumer consumer;
+
+    private JmsRecord currentRecord;
+    private Instant currentTimestamp;
+
+    public UnboundedJmsReader(
+        UnboundedJmsSource source,
+        @Nullable JmsCheckpointMark checkpointMark) {
+      this.source = source;
+      this.checkpointMark = checkpointMark;
+      this.currentRecord = null;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory = source.connectionFactory;
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        if (source.topic != null) {
+          this.consumer = this.session.createConsumer(this.session.createTopic(source.topic));
+          this.destination = source.topic;
+        } else {
+          this.consumer = this.session.createConsumer(this.session.createQueue(source.queue));
+          this.destination = source.queue;
+        }
+
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        TextMessage message = (TextMessage) this.consumer.receiveNoWait();
+
+        if (message == null) {
+          return false;
+        }
+
+        Map<String, Object> headers = new HashMap<>();
```

Comment:
By the way, as the build worked on my machine using `clean install -Papache-release` and `verify`, it would be great to include dependencies check in the release profile by default. Else, it means that an user may not be able to predict a Jenkins build issue.